beps/00010-event-auditor/README.md

```diff
@@ -0,0 +1,221 @@
+
+```ts
+export type AuditRequest = {
+  body: any;
+  url: string;
+  method: string;
+  params?: any;
+  query?: any;
+};
+
+export type AuditResponse = {
+  status: number;
+  body?: any;
+};
+```
+
+#### Audit Event Status Interfaces
+
+These interfaces define the possible statuses for an audit event entry. There are three options:
+
+```ts
+/**
+ * Indicates the event was successful.
+ */
+export type AuditEventSuccessStatus = { status: 'succeeded' };
+
+/**
+ * Indicates the event failed and includes details about the encountered errors.
+ */
+export type AuditEventFailureStatus = {
+  status: 'failed';
+  errors: ErrorLike[];
+};
+
+/**
+ * Indicates the event failed with errors that don't perfectly fit the expected structure.
+ */
+export type AuditEventUnknownFailureStatus = {
+  status: 'failed';
+  errors: unknown[];
+};
+
+export type AuditEventStatus = AuditEventSuccessStatus | AuditEventFailureStatus;
+```
+
+#### EventAuditor Interface
+
+This interface defines the functionalities of an `EventAuditor` class. This class provides methods for:
+- Extracting the actor ID from an Express request (if available).
+- Creating detailed audit event information based on provided options.
+- Logging an audit event with a specific level (info, debug, warn, or error).
+
+```ts
+/**
+ * Common fields of an audit event. Note: timestamp and pluginId are automatically added at event creation.
+ *
+ * @public
+ */
+export type AuditEventDetails = {
+  actor: ActorDetails;
+  eventName: string;
+  stage: string;
+  request?: AuditRequest;
+  response?: AuditResponse;
+  meta: JsonValue;
+  isAuditEvent: true;
+} & AuditEventStatus;
+
+export type AuditEventDetailsOptions = {
+  eventName: string;
+  stage: string;
+  metadata?: JsonValue;
+  response?: AuditResponse;
+  actorId?: string;
+  request?: Request;
+} & (AuditEventSuccessStatus | AuditEventUnknownFailureStatus);
+
+export type AuditEventOptions = {
+  eventName: string;
+  message: string;
+  stage: string;
+  level?: 'info' | 'debug' | 'warn' | 'error';
+  actorId?: string;
+  metadata?: JsonValue;
+  response?: AuditResponse;
+  request?: Request;
+} & (AuditEventSuccessStatus | AuditEventUnknownFailureStatus);
+
+export type EventAuditorOptions = {
+  logger: LoggerService;
+  authService: AuthService;
+  httpAuthService: HttpAuthService;
+};
+
+export interface EventAuditor {
+  /**
+   * Processes an express request and obtains the actorId from it. Returns undefined if actorId is not obtainable.
+   *
+   * @public
+   */
+  getActorId(request?: Request): Promise<string | undefined>;
+
+  /**
+   * Generates the audit event details to place in the metadata argument of the logger
+   *
+   * Secrets in the metadata field and request body, params and query field should be redacted by the user before passing in the request object
+   * @public
+   */
+  createAuditEventDetails(
```

Comment:
We don't really need this to be a method on the service interface right? I'm thinking this could just be part of the underlying implementation instead, and how it might call the logger. There's no need to plugin code to be calling this method right?

Reply:
It was initially just included in case users wanted to construct a custom audit log to supply to the `meta` field of the logger.